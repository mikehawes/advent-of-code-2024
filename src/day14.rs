mod robot;
pub mod robots;
