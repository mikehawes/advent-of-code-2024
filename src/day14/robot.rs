pub struct Robot {
    position: Position,
    velocity: Velocity,
}

pub type Position = [usize; 2];

pub type Velocity = [usize; 2];

pub type FloorSize = [usize; 2];

impl Robot {
    pub fn parse(_: &str) -> Robot {
        Robot {
            position: [0, 0],
            velocity: [0, 0],
        }
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;
    use std::collections::HashMap;

    pub fn print(robots: &Vec<Robot>, floor: FloorSize) -> String {
        let mut position_to_count: HashMap<Position, usize> = HashMap::new();
        for robot in robots {
            position_to_count
                .entry(robot.position)
                .and_modify(|count| *count += 1)
                .or_insert(1);
        }
        let mut str = String::new();
        for y in 0..floor[1] {
            for x in 0..floor[0] {
                str.push_str(
                    &position_to_count
                        .get(&[x, y])
                        .map(|c| c.to_string())
                        .unwrap_or(".".to_string()),
                );
            }
            str.push('\n');
        }
        str
    }
}
