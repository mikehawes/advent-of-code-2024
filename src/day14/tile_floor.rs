pub struct TileFloor {}

impl TileFloor {
    pub fn width_height(_: usize, _: usize) -> TileFloor {
        TileFloor {}
    }
}
