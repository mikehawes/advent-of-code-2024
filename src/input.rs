use std::fs::read_to_string;
use std::io;
use std::path::{Path, PathBuf};

pub fn input_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
    let full_path = file_in_src(path);
    read_to_string(full_path)
}

pub fn file_in_src<P: AsRef<Path>>(path: P) -> PathBuf {
    let mut full_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    full_path.push("src");
    full_path.push(path);
    full_path
}
