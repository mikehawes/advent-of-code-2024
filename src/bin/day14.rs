use advent_of_code_2024::day14::robots::Robots;
use advent_of_code_2024::input::input_to_string;

fn main() {
    let input = input_to_string("day14/input.txt").unwrap();
    let robots = Robots::parse(input.as_str());
    let floor = [101, 103];
    println!("Part 1: {}", robots.safety_factor_after_seconds(100, floor));
}
