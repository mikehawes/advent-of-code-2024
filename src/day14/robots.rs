use crate::day14::robot::{FloorSize, Robot};

pub struct Robots {
    robots: Vec<Robot>,
}

impl Robots {
    pub fn parse(string: &str) -> Robots {
        let robots = string.lines().map(Robot::parse).collect();
        Robots { robots }
    }
    pub fn safety_factor_after_seconds(&self, _: usize, _: FloorSize) -> usize {
        0
    }
}

#[cfg(test)]
mod tests {
    use crate::day14::robot;
    use crate::day14::robot::FloorSize;
    use crate::day14::robots::Robots;
    use crate::input::input_to_string;
    use insta::assert_snapshot;

    #[test]
    fn can_print_robots() {
        let string = input_to_string("day14/example.txt").unwrap();
        let robots = Robots::parse(&string);
        assert_snapshot!(print(&robots, [11, 7]))
    }

    fn print(robots: &Robots, floor: FloorSize) -> String {
        robot::tests::print(&robots.robots, floor)
    }
}
