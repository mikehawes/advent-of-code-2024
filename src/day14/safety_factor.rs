use crate::day14::robot::{FloorSize, Robot};

pub fn safety_factor(robots: &[Robot], floor: FloorSize) -> usize {
    let robots_top_left = robots
        .iter()
        .filter(|r| r.before_mid(floor, 0))
        .filter(|r| r.before_mid(floor, 1))
        .count();
    let robots_top_right = robots
        .iter()
        .filter(|r| r.after_mid(floor, 0))
        .filter(|r| r.before_mid(floor, 1))
        .count();
    let robots_bottom_left = robots
        .iter()
        .filter(|r| r.before_mid(floor, 0))
        .filter(|r| r.after_mid(floor, 1))
        .count();
    let robots_bottom_right = robots
        .iter()
        .filter(|r| r.after_mid(floor, 0))
        .filter(|r| r.after_mid(floor, 1))
        .count();
    robots_top_left * robots_top_right * robots_bottom_left * robots_bottom_right
}

#[cfg(test)]
mod tests {
    use crate::day14::robot::{move_for_seconds, Robot};
    use crate::day14::safety_factor::safety_factor;
    use crate::input::input_to_string;

    #[test]
    fn can_find_safety_factor() {
        let string = input_to_string("day14/example.txt").unwrap();
        let floor = [11, 7];
        let robots = move_for_seconds(&Robot::parse_vec(&string), floor, 100);
        assert_eq!(safety_factor(&robots, floor), 12)
    }
}
