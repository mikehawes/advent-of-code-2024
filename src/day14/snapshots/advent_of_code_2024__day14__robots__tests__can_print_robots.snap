---
source: src/day14/robots.rs
assertion_line: 29
expression: "print(&robots, [11, 7])"
snapshot_kind: text
---
12..........
...........
...........
...........
...........
...........
...........
